Issue #1762: Pipelines should have one source of truth

rturrado opened the pull request:
**Context:**
Currently we have two pipeline registries: [the frontend pipeline](https://github.com/PennyLaneAI/catalyst/blob/main/frontend/catalyst/pipelines.py) is used by default qjit program flow, and [the CPP pipeline](https://github.com/PennyLaneAI/catalyst/blob/main/mlir/lib/Driver/Pipelines.cpp) is used by CLI and things like having stage names in the pipeline being passed into the `pipeline` kwarg on qjit, e.g. in the oqd, mbqc, ppm pytests.

This seems redundant, not to mention error-prone when developing and maintaining.

With more and more custom pipelines emerging and people start thinking about pipeline organization more (e.g. [#1758](https://github.com/PennyLaneAI/catalyst/pull/1758)), it would be good to have pipeline registration in just one place.

**Description of the Change:**
- Set `pipelineList` in `mlir/include/DefaultPipelines.h` as the only source of truth for pipeline and pass names.
  This pipeline list is a vector of `PipelineInfo`, each of which is a pair of pipeline name and pass names.
  The pass names are just vectors of strings too.
- Update `mlir/lib/Driver/Pipelines.cpp` and `frontend/catalyst/pipelines.py` to use `DefaultPipelines.h`, via a few accessor functions.
  These were the 2 points where information was being duplicated at the moment.
  In order for the Python code to use the C++ code,  add a `default_pipelines` nanobind module to `frontend/catalyst`.

**Benefits:**
The definition of the pipelines and passes, used by C++ and Python code, is defined just in one point.

**Possible Drawbacks:**
- We introduce `mlir/include/Driver/DefaultPipelines.h` as a dependency to the `frontend`. It's not a big deal but it is ugly to depend on files outside the project.
- I don't think this PR solves the two pipeline registration issue. It just proposes a solution for the duplication of data. But I fear that the issue complains about the duplication of the processes of registration too.
- On top of this last point, I think that the CPP pipeline code could be refactored a lot. I assume this is going to be accomplished in a near future work.

**Related GitHub Issues:**
#1762

rturrado committed:
Initial commit

Set pipelineList in mlir/include/DefaultPipelines.h as the only source of truth for pipeline and pass names.
Update mlir/lib/Driver/Pipelines.cpp to use getEnforceRuntimeInvariantsPipeline() function from DefaultPipelines.h.
Update frontend/catalyst/pipelines.py to use get_enforce_runtime_invariants_stage() function from default_pipelines nanobind module.
Add fmt library dependency to mlir/lib/Driver.

Add default_pipelines nanobind module to frontend/catalyst.
At the moment, it only exports a couple of functions, one to get the pipeline names, and the other to get the pass names of the EnforceRuntimeInvariants stage.
Move nanobind configuration code from frontend/catalyst/utils to frontend.

Add catalyst.default_pipelines as a CMakeExtension to setup.py.

rturrado committed:
Remove cmake_minimum_required, project, and language settings from frontend/ctalyst/utils/CMakeLists.txt

rturrado committed:
Fully implement default_pipelines nanobind module

Update Pipelines.cpp and pipelines.py to use the new accessor functions.

rturrado commented:
@dime10 yes, yes, thanks!

rturrado committed:
Fix CodeFactor errors

Remove unused imports.

rturrado committed:
Run make format

rturrado committed:
Refactor create functions at Pipelines.cpp

rturrado committed:
Update fmt version

rturrado committed:
Git merge to upstream/main

rturrado committed:
Run make format

rturrado commented:
Sorry. I closed the PR by error. I have been rebasing my 1762 branch to upstream/main. I still have to update it to add the PBC passes. Then accomplish the requested changes in the review.

rturrado committed:
Update DefaultPipelines.h to resemble current upstream's Pipelines.cpp

rturrado committed:
Run clang-format

rturrado committed:
Add 'inline' pass back to DefaultPipelines.h (commented out)

rturrado committed:
Reformat comments so that clang-format doesn't align pass names with comments

rturrado committed:
Run clang-format

rturrado commented:
> Sorry. I closed the PR by error. I have been rebasing my 1762 branch to upstream/main. I still have to update it to add the PBC passes. Then accomplish the requested changes in the review.

Current status for this PR:
- Rebased to upstream/main.
- Contains the new PBC passes.
- TODO: move the the nanobind module from the frontend to the compiler.

I have found some inconsistencies between `pipelines.py` and `Pipelines.cpp`, the current definitions of pipelines and passes in the `frontend` and `catalyst`, respectively:
- For `bufferization-pipeline`:
  - `eliminate-empty-tensors` and `promote-buffers-to-stack` passes are disabled in the `pipelines.py` and enabled in `Pipelines.cpp`. At the moment, I have left them enabled in `DefaultPipelines.h`.
- For `llvm-dialect-lowering-pipeline`:
  - `qnode-to-async-lowering`, `async-func-to-async-runtime`, `async-to-async-runtime`, and `convert-async-to-llvm` passes can be enabled in `pipelines.py` when an `async_qnodes` flag is true. Those passes are not present in `Pipelines.cpp`. `DefaultPipelines.h` allows adding them by passing a true `asyncQNodes` flag to `getLLVMDialectLoweringStage`.
  - `add-exception-handling` pass can be enabled in `pipelines.py` when an `async_qnodes` flag is true. That pass is always enabled in `Pipelines.cpp`. `DefaultPipelines.h` allows adding it by passing a true `asyncQNodes` flag to `getLLVMDialectLoweringStage`.
  - `convert-catalyst-to-llvm` pass goes before `convert-pbc-to-llvm` pass in `default_pipelines.py`, while the order is the opposite in `Pipelines.cpp`. `DefaultPipelines.h` uses the `pipelines.py` order. Otherwise, a lot of errors like the one below are raised when running the tests, making all MLIR to LLVM lowering tests fail:

```
grad.mul:26:3: error: failed to legalize operation 'gradient.custom_grad' that was explicitly marked illegal
  "gradient.custom_grad"() {callee = @callback_matrix_multiply_callback_125154215013664, forward = @matrix_multiply_fwd.fwd, reverse = @matrix_multiply_bwd.rev} : () -> ()
  ^
grad.mul:26:3: note: see current operation: "gradient.custom_grad"() <{callee = @callback_matrix_multiply_callback_125154215013664, forward = @matrix_multiply_fwd.fwd, reverse = @matrix_multiply_bwd.rev}> : () -> ()
While processing 'CatalystConversionPass' pass Failed to lower MLIR module
```

rturrado committed:
Revisit DefaultPipelines.h
Disable clang format for pipelineList so that passes can be written one per line.
Add qnode-to-async-lowering, async-func-to-async-runtime, async-to-async-runtime and convert-async-to-llvm back to llvm-dialect-lowering-pipeline.
Add 'Disabled pass' comments to explicitly mark disabled passes.
Remove InlinePass from BufferizationPipeline as it is not present neither in pipelines.py nor in Pipelines.cpp.

rturrado committed:
Remove unroll-conditional-ppr-ppm pass from quantum-compilation-pipeline

rturrado committed:
Move default_pipelines nanobind module from frontend to catalyst (WIP)
Remove default_pipelines nanobind module code from frontend/catalyst/CMakeLists.txt.
Add default_pipelines nanobind module code to mlir/lib/Driver/CMakeLists.txt.
Move frontend/catalyst/default_pipelines.cpp to mlir/lib/Driver/DefaultPipelines.cpp.
Update setup.py to get catalyst.default_pipelines extension from mlir/lib/Driver.

rturrado committed:
Move default_pipelines nanobind module from frontend to catalyst
Create DefaultPipelines folder both under mlib/include/Driver and mlir/lib/Driver.
Move DefaultPipelines.h to mlir/include/Driver and DefaultPipelines.cpp to mlir/lib/Driver.
Move default_pipelines nanobind module CMake code to mlir/include/Driver/CMakeLists.txt.
Update target include directories of default_pipelines nanobind module to point to mlir/include/Driver.
Remove linkage of default_pipelines nanobind module against CatalystCompilerDriver.
Update setup.py to build extensions on different folders.
Update setup.py to create catalyst.default_pipelines extension from mlir/lib/Driver/DefaultPipelines folder.

rturrado commented:
@dime10 I've now moved the `default_pipelines` nanobind module from the frontend to the compiler. The code is passing `make test` locally. However, I'm still not sure if this is what you had in mind.

- There is now a `DefaultPipelines` subfolder both within `mlir/include/Driver` and `mlir/lib/Driver`.

- `mlir/include/Driver/DefaultPipelines` holds:
  - `DefaultPipelines.h`, the new one-source-of-truth. This file is used both from the frontend (through the `default_pipelines` nanobind module) and from the compiler (through the Driver code). **Note** however that the `DefaultPipelines.h` doesn't have any dependencies with the rest of the Catalyst code. This was already intended in the original implementation to minimise coupling between the frontend and the compiler.

- `mlir/lib/Driver/DefaultPipelines` holds:
  - `DefaultPipelines.cpp` with the `NB_MODULE` code, and a
  - `CMakeLists.txt` with a `project(DefaultPipelines)` definition. This is needed when `cmake` is invoked from `setup.py` using `mlir/lib/Driver/DefaultPipelines` as root folder. **Note** that, if we were to use `mlir` as root folder from `setup.py`, the whole of the compiler would be compiled; while the nanobind module is really needed.

dime10 commented:
Thanks for discovering (and fixing!) those inconsistencies in the default pipeline! In general I would say `pipelines.py` is probably the more up-to-date one, because that's what we use more often. Let me have a look in detail:

> - eliminate-empty-tensors and promote-buffers-to-stack passes are disabled in the pipelines.py and enabled in Pipelines.cpp. At the moment, I have left them enabled in DefaultPipelines.h.

I would say we should comment them out in `DefaultPipelines.h` as well, with the [same](https://github.com/PennyLaneAI/catalyst/blob/41b3e6e701f3f6b24962f514136f31b152b2355d/frontend/catalyst/pipelines.py#L319-L321) [notes](https://github.com/PennyLaneAI/catalyst/blob/41b3e6e701f3f6b24962f514136f31b152b2355d/frontend/catalyst/pipelines.py#L305-L308) from the frontend.

> - For llvm-dialect-lowering-pipeline:

Everything looks good to me here :)

rturrado committed:
Add catalyst.default_pipelines extension to .pylintrc

rturrado committed:
Disable eliminate-empty-tensors and promote-buffers-to-stack passes from bufferization-pipeline

rturrado commented:
> > * eliminate-empty-tensors and promote-buffers-to-stack passes are disabled in the pipelines.py and enabled in Pipelines.cpp. At the moment, I have left them enabled in DefaultPipelines.h.
> 
> I would say we should comment them out in `DefaultPipelines.h` as well, with the [same](https://github.com/PennyLaneAI/catalyst/blob/41b3e6e701f3f6b24962f514136f31b152b2355d/frontend/catalyst/pipelines.py#L319-L321) [notes](https://github.com/PennyLaneAI/catalyst/blob/41b3e6e701f3f6b24962f514136f31b152b2355d/frontend/catalyst/pipelines.py#L305-L308) from the frontend.

Perfect, thanks! I'll address this in the following commits.

I will also revert all the changes in the frontend because this PR shouldn't really affect the frontend, should it?

And I think there may be room for a couple of follow-up PRs, but I will leave that completely to you:
1. `setup.py` is now invoking `cmake` from `frontend` to build the `wrapper` extension, but I think it should suffice to invoke it from `frontend/catalyst/utils`. However, that would probably change the way `wrapper` is imported (e.g., from something like `from catalyst.utils import wrapper` to `from catalyst import wrapper`). And I don't know if you would be happy with that change.
2. I would also revisit the names of the passes. At the moment, you can find a mix of verbs and nouns, e.g., `lower-mitigation` and `adjoint-lowering`, which, in my opinion, calls for more consistency.

rturrado committed:
Revert frontend changes

dime10 commented:
Thanks @rturrado, I was actually in the process of testing out some changes, in particular I'm trying not to invoke the DefaultPipelines CMake script from setup.py but directly in the MLIR build. The nanobind module would then be discovered in the same way the compiler CLI and other artifacts are from the `mlir` and `runtime` systems. I can open a patch PR against your branch, or just open a follow up PR if you prefer :)

Overall your PR looks amazing btw 🎉

rturrado commented:
> Thanks @rturrado, I was actually in the process of testing out some changes, in particular I'm trying not to invoke the DefaultPipelines CMake script from setup.py but directly in the MLIR build. The nanobind module would then be discovered in the same way the compiler CLI and other artifacts are from the `mlir` and `runtime` systems.

Ah, yes, good. That is something I hadn't fully understood from your comments above. I should have asked for clarification. Anyway, I am very glad you are addressing it now.

> I can open a patch PR against your branch, or just open a follow up PR if you prefer :)

I leave that up to you. Whatever you decide is best.

> Overall your PR looks amazing btw 🎉

Many thanks! Appreciated.

dime10 committed:
Move extension module build step to mlir

To avoid build-time links between the three primary components of
Catalyst (runtime, mlir, frontend), the task of building the default
pipelines extension module (which is defined in /mlir) is relegated to
the mlir build step. Similar to other libraries (e.g. the runtime
callback registry), the pipelines module is discovered by the frontend
at run-time.

dime10 committed:
Merge branch 'main' into 1762

dime10 commented on .pylintrc:
```diff
@@ -2,7 +2,7 @@
 # be loaded. Extensions are loading into the active Python interpreter and may
 # run arbitrary code
-extension-pkg-allow-list=catalyst.utils.wrapper
+extension-pkg-allow-list=catalyst.utils.wrapper,catalyst.default_pipelines
```
We might need to update this? I don't remember exactly how this setting works

rturrado replied:
I had added `catalyst.default_pipelines` after having added it to the list of Python extensions in `setup.py`. With the current code, it may be not necessary, but I have to check it.

rturrado replied:
I've just removed it. Because `default_pipelines` is discovered at runtime, it wouldn't be locatable at linting time.

dime10 reviewed (approved):
There was another pass added on main that wasn't added to the C++ definition (see merge commit), so it's a good thing this PR is finally going in :) 

I think we can ignore codefactor, this pr doesn't modify anything in the flagged files 🤔 

Changelog should be added though, this is probably an internal change because the `pipelines.py` still works the same for users.

EDIT: Ah, looks like we need to ensure nanobind is available for the mlir build step in CI, makes sense!

dime10 committed:
Fix Python discovery in wheels

dime10 committed:
silence import error coverage

dime10 committed:
silence nanobind warnings

dime10 committed:
change error to warning

dime10 commented:
Alright looks like we're good to go, the last failure from the mac wheels is coming from main 👌

rturrado commented:
> There was another pass added on main that wasn't added to the C++ definition (see merge commit), so it's a good thing this PR is finally going in :)

Great!

> I think we can ignore codefactor, this pr doesn't modify anything in the flagged files 🤔

OK.

> Changelog should be added though, this is probably an internal change because the `pipelines.py` still works the same for users.

I can add that if you haven't already.

rturrado committed:
Update changelog-dev.md

rturrado committed:
Remove catalyst.default_pipelines from .pylintrc

The default_pipelines nanobind module is discovered at runtime by appending
the Catalyst lib directory to sys.path, not installed as a Python package via setup.py. The extension-pkg-allow-list entry was therefore incorrect and had no effect.

rturrado committed:
Merge branch 'main' into 1762

mehrdad2m reviewed (approved):
Looks Great Thanks @rturrado!

rturrado commented:
> Looks Great Thanks @rturrado!

Thanks to you guys, a pleasure!
